### ISSUE TITLE
TypeError with to_file() to GPKG or ESRI Shapefile

### ISSUE THREAD

**Issue by lsetiawan** [geopandas/geopandas#353]
Hi,

When I try to write my GeoDataFrame to Geopackage or Shapefile, I am getting TypeError. I have been struggling with this for a couple of days. I'm not sure if this is directly a geopandas issue or fiona issue. Thanks.

FYI, I am using geopandas 0.2, python 2.7 from channel conda-forge

Here's what I tried
``` python
out_pth = '/home/lsetiawan/Desktop'
gdf.to_file(os.path.join(out_pth,'nanoos_nvs.gpkg'), driver='GPKG')
```
or 
``` python
out_pth = '/home/lsetiawan/Desktop'
gdf.to_file(os.path.join(out_pth,'nanoos_nvs.shp'), driver='ESRI Shapefile')
```
This is the TypeError that I get (Same thing if I try to output ESRI Shapefile)
```
TypeError                                 Traceback (most recent call last)
<ipython-input-25-7f7493d73b63> in <module>()
      1 out_pth = '/home/lsetiawan/Desktop'
----> 2 gdf.to_file(os.path.join(out_pth,'nanoos_nvs.gpkg'), driver='GPKG')

/home/lsetiawan/anaconda2/envs/geo/lib/python2.7/site-packages/geopandas/geodataframe.pyc in to_file(self, filename, driver, schema, **kwargs)
    341         """
    342         from geopandas.io.file import to_file
--> 343         to_file(self, filename, driver, schema, **kwargs)
    344 
    345     def to_crs(self, crs=None, epsg=None, inplace=False):

/home/lsetiawan/anaconda2/envs/geo/lib/python2.7/site-packages/geopandas/io/file.pyc in to_file(df, filename, driver, schema, **kwargs)
     61                     schema=schema, **kwargs) as c:
     62         for feature in df.iterfeatures():
---> 63             c.write(feature)
     64 
     65 

/home/lsetiawan/anaconda2/envs/geo/lib/python2.7/site-packages/fiona/collection.pyc in write(self, record)
    335     def write(self, record):
    336         """Stages a record for writing to disk."""
--> 337         self.writerecords([record])
    338 
    339     def validate_record(self, record):

/home/lsetiawan/anaconda2/envs/geo/lib/python2.7/site-packages/fiona/collection.pyc in writerecords(self, records)
    329         if self.mode not in ('a', 'w'):
    330             raise IOError("collection not open for writing")
--> 331         self.session.writerecs(records, self)
    332         self._len = self.session.get_length()
    333         self._bounds = self.session.get_extent()

fiona/ogrext.pyx in fiona.ogrext.WritingSession.writerecs (fiona/ogrext.c:17078)()

fiona/ogrext.pyx in fiona.ogrext.OGRFeatureBuilder.build (fiona/ogrext.c:6587)()

TypeError: function takes exactly 5 arguments (1 given)
```

**Comment by jorisvandenbossche** [comment]
cc @sgillies Do you know if this is a geopandas or fiona issue?

**Comment by micahcochran** [comment]
@lsetiawan What version of Fiona is installed?  Is it using GDAL version 2.0.x?

**Comment by lsetiawan** [comment]
@micahcochran This is what I got:
- numpy	1.11.1
- pandas	0.18.1
- geopandas	0.2
- fiona	1.6.4
- shapely	1.5.16
- gdal	1.11.4

Thanks for your help!

**Comment by micahcochran** [comment]
@lsetiawan There is simply not enough information posted for anyone to be able to replicate the problem.    We need atleast a small snippet of your code that has the problem or the a snippet of data that shows the problem.

Try some simple tests to verify if Fiona is working property with a shapefile on the command line:
```
$ fio dump in.shp | fio load outof.shp --driver Shapefile
```
This copies the shapefile using fiona and reprojects it into WGS 84 (without writing the .PRJ file).

Try a simple loop test with geopandas too.

Without knowing what version of Cython was used, there isn't really enough information to be able to easily where it is going wrong in Fiona.  It would be really hard to figure out the offending line of Cython code.

I am slightly leaning toward this Fiona problem, but I really don't have much information to go on.

**Comment by lsetiawan** [comment]
@micahcochran Thank you for your answer. I have tried the test command and ran a simple loot with geopandas, and it appears that Fiona is working properly. I was able to copy the shapefile. I tried the same process with Geopackage, and it is not working. 

Here is a link to the ipython notebook of the problem: [CSV2GPKG](https://github.com/lsetiawan/stuff/blob/master/CSV_To_GPKG%20Layer.ipynb)
Then here's a link to the csv data that I'm using: [CSV_Data](https://github.com/lsetiawan/stuff/blob/master/platform_meta_201606291445_1.csv)

I am also including the exact environment I'm using:
- backports=1.0=py27_0
- backports.shutil_get_terminal_size=1.0.0=py27_0
- backports_abc=0.4=py27_0
- click=6.6=py27_0
- click-plugins=1.0.3=py27_0
- cligj=0.4.0=py27_0
- configparser=3.5.0b2=py27_1
- curl=7.49.1=0
- cycler=0.10.0=py27_0
- decorator=4.0.10=py27_0
- descartes=1.0.2=py27_0
- entrypoints=0.2.2=py27_0
- expat=2.1.0=1
- fiona=1.6.4=np111py27_1
- fontconfig=2.11.1=3
- freetype=2.6.3=1
- freexl=1.0.2=1
- functools32=3.2.3.2=py27_1
- gdal=1.11.4=np111py27_10
- geopandas=0.2=py27_0
- geopy=1.10.0=py27_0
- geos=3.4.2=4
- giflib=5.1.2=1
- hdf4=4.2.11=4
- hdf5=1.8.17=2
- icu=56.1=4
- ipykernel=4.3.1=py27_0
- ipython=4.2.0=py27_2
- ipython_genutils=0.1.0=py27_0
- ipywidgets=5.1.5=py27_0
- jasper=1.900.1=5
- jinja2=2.8=py27_1
- jpeg=9b=0
- json-c=0.12=0
- jsonschema=2.5.1=py27_0
- jupyter=1.0.0=py27_3
- jupyter_client=4.3.0=py27_0
- jupyter_console=4.1.1=py27_0
- jupyter_core=4.1.0=py27_0
- krb5=1.14.2=0
- libgfortran=3.0.0=1
- libiconv=1.14=2
- libnetcdf=4.4.0=1
- libpng=1.6.23=0
- libsodium=1.0.10=0
- libspatialindex=1.8.5=1
- libspatialite=4.3.0a=8
- libtiff=4.0.6=5
- libxml2=2.9.3=8
- markupsafe=0.23=py27_2
- matplotlib=1.5.1=np111py27_4
- mistune=0.7.3=py27_0
- mkl=11.3.3=0
- munch=2.0.4=py27_0
- nbconvert=4.2.0=py27_0
- nbformat=4.0.1=py27_0
- notebook=4.2.1=py27_0
- numpy=1.11.1=py27_0
- openjpeg=2.1.0=5
- openssl=1.0.2h=1
- pandas=0.18.1=np111py27_0
- path.py=8.2.1=py27_0
- pathlib2=2.1.0=py27_0
- pexpect=4.1.0=py27_0
- pickleshare=0.7.2=py27_0
- pip=8.1.2=py27_0
- postgresql=9.1.4=1
- proj.4=4.9.2=0
- psycopg2=2.6.1=py27_1
- ptyprocess=0.5.1=py27_0
- pygments=2.1.3=py27_0
- pyparsing=2.1.4=py27_0
- pyproj=1.9.5.1=py27_0
- pyqt=4.11.4=py27_3
- pysal=1.11.1=py27_0
- python=2.7.12=0
- python-dateutil=2.5.3=py27_0
- python-decorator=4.0.10=py27_0
- python-pathlib2=2.1.0=py27_0
- python-simplegeneric=0.8.1=py27_0
- pytz=2016.4=py27_0
- pyzmq=15.2.0=py27_1
- qt=4.8.7=3
- qtconsole=4.2.1=py27_0
- readline=6.2=2
- rtree=0.8.2=py27_2
- scipy=0.17.1=np111py27_1
- setuptools=23.0.0=py27_0
- shapely=1.5.16=np111py27_0
- singledispatch=3.4.0.3=py27_0
- sip=4.18=py27_0
- six=1.10.0=py27_0
- sqlalchemy=1.0.13=py27_0
- sqlite=3.13.0=1
- ssl_match_hostname=3.5.0.1=py27_0
- terminado=0.6=py27_0
- tk=8.5.18=0
- tornado=4.3=py27_1
- traitlets=4.2.1=py27_0
- version_information=1.0.3=py27_0
- wheel=0.29.0=py27_0
- widgetsnbextension=1.2.3=py27_1
- xerces-c=3.1.2=0
- xz=5.0.5=1
- zeromq=4.1.5=0
- zlib=1.2.8=3
- pip:
  - backports-abc==0.4
  - backports.shutil-get-terminal-size==1.0.0
  - backports.ssl-match-hostname==3.5.0.1
  - ipython-genutils==0.1.0
  - jupyter-client==4.3.0
  - jupyter-console==4.1.1
  - jupyter-core==4.1.0
  - simplegeneric==0.8.1
  - version-information==1.0.3

I hope that I've provided enough information. Thanks.

**Comment by micahcochran** [comment]
I poked around on Anaconda's site and I assume it is [Cython 0.24](http://conda.anaconda.org/conda-forge/osx-64). 

GDAL has to be compiled with libsqlite3.  If GDAL has geopackage support Geopackages it will report this:
```
$ ogr2ogr --formats | grep GPKG
  -> "GPKG" (read/write)
```

**Comment by lsetiawan** [comment]
@micahcochran I have performed `$ ogr2ogr --formats | grep GPKG` and the result is the exact result that you have. 
```
(geo)lsetiawan@macos:~/Desktop/goaon_assets$ ogr2ogr --formats | grep GPKG
  -> "GPKG" (read/write)
```

**Comment by sgillies** [comment]
I'm almost certain this is the bug I fixed (for 1.7.0) with https://github.com/Toblerity/Fiona/issues/356. The fix has not yet been ported back to Fiona 1.6.x.

The workaround: 

1. When you open a file in 'w' mode, specify an `encoding` keyword argument like 'latin1' or 'gb18030' or Fiona will make a guess at the appropriate encoding (potentially a wrong guess).
2. All string properties of the features you write to the opened file must have Python unicode values. They are encoded to utf-8 before being passed to GDAL functions. If they're not unicode, attempts to encode them will fail.

The gist of the bug I referenced in this comment: I had bungled the re-raising of unicode errors from 1 and 2.

**Comment by lsetiawan** [comment]
@sgillies Thank you for your workaround. I was able to run the code with the workaround. It worked once, then I try running it again, saving as a different file, but now I'm getting a RuntimeError:
``` python
RuntimeError                              Traceback (most recent call last)
<ipython-input-20-97c90042f082> in <module>()
----> 1 createGPKG(feat_list,schema,crs)

<ipython-input-17-a8f36309cdb3> in createGPKG(features, schema, crs)
      8         for num in range(0,len(features)):
      9             print features[num]['id']
---> 10             c.write(features[num])

/home/lsetiawan/anaconda2/envs/geo/lib/python2.7/site-packages/fiona/collection.pyc in write(self, record)
    335     def write(self, record):
    336         """Stages a record for writing to disk."""
--> 337         self.writerecords([record])
    338 
    339     def validate_record(self, record):

/home/lsetiawan/anaconda2/envs/geo/lib/python2.7/site-packages/fiona/collection.pyc in writerecords(self, records)
    329         if self.mode not in ('a', 'w'):
    330             raise IOError("collection not open for writing")
--> 331         self.session.writerecs(records, self)
    332         self._len = self.session.get_length()
    333         self._bounds = self.session.get_extent()

fiona/ogrext.pyx in fiona.ogrext.WritingSession.writerecs (fiona/ogrext.c:17128)()

RuntimeError: Failed to write record: {u'geometry': {u'type': u'Point', u'coordinates': [-123.774, 46.2047]}, u'type': u'Feature', u'id': u'13', u'properties': {u'platform_label': u'CMOP_Coaof', u'provider_url': u'http://www.stccmop.org', u'data_source': u'NANOOS-CMOP', u'name': u'Waste water outfall (City of Astoria)', u'short_name': None, u'url': u'http://www.stccmop.org/datamart/observation_network/fixedstation?station=coaof', u'datasrc_platform_cd': u'coaof', u'status_date': 0, u'lon': -123.774, u'sub_region': None, u'platform_type': u'Fixed Shore Platform', u'state': u'Oregon', u'status': u'offline-temporary', u'data_source_url': None, u'bounding_area_hue': 0, u'provider': u'CMOP', u'lat': 46.2047, u'icon': u'shore_station_default', u'data_timezone': u'US/Pacific', u'region': u'Columbia Estuary', u'provider_type': u'Academic'}}
```
Here's the function and data:
``` python
# Create GPKG using only fiona
def createGPKG(features,schema,crs):
    out_pth = './'
    
    with fio.open(os.path.join(out_pth,'nanoos_nvs66.gpkg'), 'w', driver="GPKG", schema = schema,crs_wkt=crs,encoding='gb18030') as c:
        for num in range(0,len(features)):
            print features[num]['id']
            c.write(features[num])
```
Features sample:
```
{u'geometry': {u'coordinates': [-124.9502, 47.9672], u'type': u'Point'},
 u'id': u'0',
 u'properties': {u'bounding_area_hue': 0,
  u'data_source': u'NANOOS-APL2',
  u'data_source_url': u'http://www.nanoos.org/',
  u'data_timezone': u'US/Pacific',
  u'datasrc_platform_cd': u'aplchaba',
  u'icon': u'buoy_default',
  u'lat': 47.9672,
  u'lon': -124.9502,
  u'name': u'?h\xe1?ba\xb7 UW/NANOOS Moored Buoy near La Push',
  u'platform_label': u'APL_Chaba',
  u'platform_type': u'Buoy',
  u'provider': u'APL-UW',
  u'provider_type': u'Academic',
  u'provider_url': u'http://www.nanoos.org/',
  u'region': u'La Push',
  u'short_name': u'APL-UW ?h\xe1?ba\xb7',
  u'state': u'Washington',
  u'status': u'online',
  u'status_date': 1458680400,
  u'sub_region': None,
  u'url': u'http://wavechasers.apl.washington.edu/projects/active-projects/nemo'},
 u'type': u'Feature'}
```

Thanks!

**Comment by lsetiawan** [comment]
Thank you all for the help, I think the issue has been solved.

**Comment by tollask** [comment]
@lsetiawan how did you resolve the RuntimeError?

**Comment by BAFurtado** [comment]
> I'm almost certain this is the bug I fixed (for 1.7.0) with [Toblerity/Fiona#356](https://github.com/Toblerity/Fiona/issues/356). The fix has not yet been ported back to Fiona 1.6.x.
> 
> The workaround:
> 
>     1. When you open a file in 'w' mode, specify an `encoding` keyword argument like 'latin1' or 'gb18030' or Fiona will make a guess at the appropriate encoding (potentially a wrong guess).
> 
>     2. All string properties of the features you write to the opened file must have Python unicode values. They are encoded to utf-8 before being passed to GDAL functions. If they're not unicode, attempts to encode them will fail.
> 
> 
> The gist of the bug I referenced in this comment: I had bungled the re-raising of unicode errors from 1 and 2.
> 
> Upgrading to Fiona 1.7.0 is another solution.

Hmmm. I got a RunTimeError when saving, using Fiona 1.8.4
  File "fiona/ogrext.pyx", line 1182, in fiona.ogrext.WritingSession.writerecs
RuntimeError: Failed to write record:

**Comment by busarieedris** [comment]
I dont think I got how this issue was fixed

**Comment by busarieedris** [comment]
> @sgillies Thank you for your workaround. I was able to run the code with the workaround. It worked once, then I try running it again, saving as a different file, but now I'm getting a RuntimeError:
> 
> ```python
> RuntimeError                              Traceback (most recent call last)
> <ipython-input-20-97c90042f082> in <module>()
> ----> 1 createGPKG(feat_list,schema,crs)
> 
> <ipython-input-17-a8f36309cdb3> in createGPKG(features, schema, crs)
>       8         for num in range(0,len(features)):
>       9             print features[num]['id']
> ---> 10             c.write(features[num])
> 
> /home/lsetiawan/anaconda2/envs/geo/lib/python2.7/site-packages/fiona/collection.pyc in write(self, record)
>     335     def write(self, record):
>     336         """Stages a record for writing to disk."""
> --> 337         self.writerecords([record])
>     338 
>     339     def validate_record(self, record):
> 
> /home/lsetiawan/anaconda2/envs/geo/lib/python2.7/site-packages/fiona/collection.pyc in writerecords(self, records)
>     329         if self.mode not in ('a', 'w'):
>     330             raise IOError("collection not open for writing")
> --> 331         self.session.writerecs(records, self)
>     332         self._len = self.session.get_length()
>     333         self._bounds = self.session.get_extent()
> 
> fiona/ogrext.pyx in fiona.ogrext.WritingSession.writerecs (fiona/ogrext.c:17128)()
> 
> RuntimeError: Failed to write record: {u'geometry': {u'type': u'Point', u'coordinates': [-123.774, 46.2047]}, u'type': u'Feature', u'id': u'13', u'properties': {u'platform_label': u'CMOP_Coaof', u'provider_url': u'http://www.stccmop.org', u'data_source': u'NANOOS-CMOP', u'name': u'Waste water outfall (City of Astoria)', u'short_name': None, u'url': u'http://www.stccmop.org/datamart/observation_network/fixedstation?station=coaof', u'datasrc_platform_cd': u'coaof', u'status_date': 0, u'lon': -123.774, u'sub_region': None, u'platform_type': u'Fixed Shore Platform', u'state': u'Oregon', u'status': u'offline-temporary', u'data_source_url': None, u'bounding_area_hue': 0, u'provider': u'CMOP', u'lat': 46.2047, u'icon': u'shore_station_default', u'data_timezone': u'US/Pacific', u'region': u'Columbia Estuary', u'provider_type': u'Academic'}}
> ```
> 
> Here's the function and data:
> 
> ```python
> # Create GPKG using only fiona
> def createGPKG(features,schema,crs):
>     out_pth = './'
> 
>     with fio.open(os.path.join(out_pth,'nanoos_nvs66.gpkg'), 'w', driver="GPKG", schema = schema,crs_wkt=crs,encoding='gb18030') as c:
>         for num in range(0,len(features)):
>             print features[num]['id']
>             c.write(features[num])
> ```
> 
> Features sample:
> 
> ```
> {u'geometry': {u'coordinates': [-124.9502, 47.9672], u'type': u'Point'},
>  u'id': u'0',
>  u'properties': {u'bounding_area_hue': 0,
>   u'data_source': u'NANOOS-APL2',
>   u'data_source_url': u'http://www.nanoos.org/',
>   u'data_timezone': u'US/Pacific',
>   u'datasrc_platform_cd': u'aplchaba',
>   u'icon': u'buoy_default',
>   u'lat': 47.9672,
>   u'lon': -124.9502,
>   u'name': u'?h\xe1?ba\xb7 UW/NANOOS Moored Buoy near La Push',
>   u'platform_label': u'APL_Chaba',
>   u'platform_type': u'Buoy',
>   u'provider': u'APL-UW',
>   u'provider_type': u'Academic',
>   u'provider_url': u'http://www.nanoos.org/',
>   u'region': u'La Push',
>   u'short_name': u'APL-UW ?h\xe1?ba\xb7',
>   u'state': u'Washington',
>   u'status': u'online',
>   u'status_date': 1458680400,
>   u'sub_region': None,
>   u'url': u'http://wavechasers.apl.washington.edu/projects/active-projects/nemo'},
>  u'type': u'Feature'}
> ```
> 
> Thanks!

I dont think I got how this particular error was fixed.Is it the Fiona upgrade?

**Comment by martinfleis** [comment]
@busarieedris Yes, it was fixed in Toblerity/Fiona#356 released in Fiona 1.7.0. If you have similar issue with the latest Fiona and latest GeoPandas, can you open a new issue for that?

**Comment by busarieedris** [comment]
I have the same issue with Fiona 2.x.x
